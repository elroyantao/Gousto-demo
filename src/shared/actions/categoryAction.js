import 'isomorphic-fetch'

const fetchCategoriesRequest = () => ({
  type: 'FETCH_CATEGORIES_REQUEST'
})

const fetchCategoriesResponse = (categories) => ({
  type: 'FETCH_CATEGORIES_RESPONSE',
  categories
})

const fetchCategoriesFailure = () => ({
  type: 'FETCH_CATEGORIES_FAILURE'
})

export const fetchCategories = () => (dispatch) => {
  dispatch(fetchCategoriesRequest())
  return fetch('/api/categories', { method: 'GET' })
  .then((res) => {
    if (!res.ok) throw Error(res.statusText)
    return res.json()
  })
  .then((res) => {
    dispatch(fetchCategoriesResponse(res.data))
  })
  .catch(() => {
    dispatch(fetchCategoriesFailure())
  })
}
