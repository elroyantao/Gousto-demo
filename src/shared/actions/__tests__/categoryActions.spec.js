import fetchMock from 'fetch-mock'
import * as actions from '../categoryAction'

const fetchAction = actions.fetchCategories()
const dispatch = jest.fn()

describe('category actions', () => {
  it('fetchCategoriesRequest()', () => {
    expect(actions.fetchCategoriesRequest()).toMatchSnapshot()
  })
  it('fetchCategoriesFailure()', () => {
    expect(actions.fetchCategoriesFailure()).toMatchSnapshot()
  })
  it('fetchCategoriesResponse()', () => {
    expect(actions.fetchCategoriesResponse([1, 2, 3])).toMatchSnapshot()
  })
  describe('fetchCategories() ', () => {
    afterEach(() => {
      fetchMock.restore()
      jest.clearAllMocks()
    })
    it('a successful fetch', () => {
      fetchMock.get('/api/categories', {
        data: [1, 2, 3]
      })
      expect.assertions(3)
      return fetchAction(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/categories', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
    it('an unsuccessful fetch', () => {
      fetchMock.get('/api/categories', 402)
      expect.assertions(3)
      return fetchAction(dispatch).then(() => {
        expect(fetchMock.lastCall()).toEqual(['/api/categories', { method: 'GET' }])
        expect(dispatch).toHaveBeenCalledTimes(2)
        expect(dispatch.mock.calls).toMatchSnapshot()
      })
    })
  })
})
