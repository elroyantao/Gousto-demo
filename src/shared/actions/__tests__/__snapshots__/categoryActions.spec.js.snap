// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`category actions fetchCategories()  a successful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_CATEGORIES_REQUEST",
    },
  ],
  Array [
    Object {
      "categories": Array [
        1,
        2,
        3,
      ],
      "type": "FETCH_CATEGORIES_RESPONSE",
    },
  ],
]
`;

exports[`category actions fetchCategories()  an unsuccessful fetch 1`] = `
Array [
  Array [
    Object {
      "type": "FETCH_CATEGORIES_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "FETCH_CATEGORIES_FAILURE",
    },
  ],
]
`;

exports[`category actions fetchCategoriesFailure() 1`] = `
Object {
  "type": "FETCH_CATEGORIES_FAILURE",
}
`;

exports[`category actions fetchCategoriesRequest() 1`] = `
Object {
  "type": "FETCH_CATEGORIES_REQUEST",
}
`;

exports[`category actions fetchCategoriesResponse() 1`] = `
Object {
  "categories": Array [
    1,
    2,
    3,
  ],
  "type": "FETCH_CATEGORIES_RESPONSE",
}
`;
